### PR TITLE
Replace underscore dependencies with amp dependencies.

### PR DESCRIPTION
**Pull request by drewloomer**
As discussed in [issue 127](https://github.com/AmpersandJS/ampersand-state/issues/127) by @jvduf . Was not able to remove the underscore dependency completely because omit has not been added to the npm registry yet. 

Update: Thanks to the modules that were published today, I was able to remove all the underscore deps.

### CODE REVIEW

**Comment by wraithgar** [comment]
@drewloomer thanks for getting this done!  Can you remove the `npm-debug.log` file from the PR?

Otherwise it's a +1 from me, checked it out locally, tests ran fine.

**Comment by drewloomer** [comment]
Oops. Sorry about that, @wraithgar . `npm-debug.log` has been removed. Should be all set now!

**Comment by HenrikJoreteg** [comment]
hey @drewloomer, while you're at it, wanna swap out `backbone-events-standalone` for `ampersand-events` then we can just get all the dependency upgrades done at once.

**Comment by drewloomer** [comment]
@HenrikJoreteg all set. should be good to go.

**Comment by wraithgar** [comment]
Yay even better. Test pass for me locally. +1

**Comment by HenrikJoreteg** [comment]
This technically adds some features, specifically [listenToAndRun](http://ampersandjs.com/docs#ampersand-events-listentoandrun) is now a method because it's using ampersand-events. So should we make this a minor release?

**Comment by wraithgar** [comment]
Yes minor release

**Comment by latentflip** [comment]
I haven't actually you know,tested anything, but conceptual +1 from me :)

File: ampersand-state.js [hunk 1]
```diff
@@ -557,14 +579,14 @@ function createPropertyDefinition(object, name, desc, isSession) {
...
+        fn: isFunction(definition) ? definition : definition.fn,
         cache: (definition.cache !== false),
         depList: definition.deps || []
     };
 
     // add to our shared dependency list
-    _.each(def.depList, function (dep) {
-        modelProto._deps[dep] = _(modelProto._deps[dep] || []).union([name]);
+    each(def.depList, function (dep) {
+        modelProto._deps[dep] = unique(flatten.call((modelProto._deps[dep] || []), true, true, []), [name]);
```
Comment (HenrikJoreteg) [inline ampersand-state.js]: Ah, we don't have a `amp-union`. I'd say we should add this amp and us it here.

File: ampersand-state.js [hunk 2]
```diff
@@ -1,13 +1,35 @@
...
-var _ = require('underscore');
-var BBEvents = require('backbone-events-standalone');
+var uniqueId = require('amp-unique-id');
+var extend = require('amp-extend');
+var escape = require('amp-escape');
+var each = require('amp-each');
+var clone = require('amp-clone');
+var bind = require('amp-bind');
+var unique = require('amp-unique');
+var flatten = require('amp-result');
```
Comment (HenrikJoreteg) [inline ampersand-state.js]: Oops, this doesn't look right.

File: ampersand-state.js [hunk 3]
```diff
@@ -587,7 +609,7 @@ var dataTypes = {
     date: {
         set: function (newVal) {
             var newType;
-            if (!_.isDate(newVal)) {
+            if (!isDate(newVal)) {
                 try {
                     var dateVal = new Date(newVal).valueOf();
                     if (isNaN(dateVal)) {
```
Comment (HenrikJoreteg) [inline ampersand-state.js]: For reasons described here: https://gist.github.com/kitcambridge/1086528 we should also switch this out to instead use `amp-is-nan`: http://amp.ampersandjs.com/#amp-is-nan

**Comment by HenrikJoreteg** [comment]
On closer review, I think we should address these few points before we publish.

File: ampersand-state.js (hunk 2 above)
Comment (drewloomer) [inline ampersand-state.js]: Happy to make a change, but I'm afraid I don't know what you mean by "doesn't look right"?

Comment (latentflip) [inline ampersand-state.js]: I assume that flatten = amp-result line is wrong :)



Philip Roberts
&yet

Comment (drewloomer) [inline ampersand-state.js]: Hah! Oops was right. Thanks, guys.

**Comment by wraithgar** [comment]
@drewloomer amp-union now exists!

**Comment by drewloomer** [comment]
Thanks for the heads up, @wraithgar . Should be all set! I wouldn't mind a double check on how I replaced Underscore's union. The original invocation was a little weird to me, and while I think I'm doing it right I'm not 100% sure.

**Comment by wraithgar** [comment]
Yeah you can simplify that. `union` guards against `undefined`, so the `|| []` is not needed, you can simply call

```
modelProto._deps[dep] = union(modelProto._deps[dep], [name]);
```

And if `_deps[dep]` doesn't exist it will simply be set to `[name]`, if it did exist it will have name added to it.

**Comment by paglias** [comment]
Any update on this?

**Comment by drewloomer** [comment]
As far as I know it's good to go, just waiting on a final blessing.

**Comment by wraithgar** [comment]
@drewloomer this needs to have merge conflicts resolved first, it would appear?  Can you pull master into your branch?

**Comment by drewloomer** [comment]
I've resolved the conflict and all the tests passed locally. Let me know if you need anything else.

**Comment by nabilfreeman** [comment]
hey guys, any plan to merge this soon?

**Comment by kamilogorek** [comment]
@nabilfreeman please bear with us, as we're still discussing some things around amp/lodash. Thanks!

**Comment by boopathi** [comment]
Looking forward to use it.
@kamilogorek  is that discussion public ? Would like to take a look at it!

**Comment by kamilogorek** [comment]
@boopathi not yet, but @HenrikJoreteg will write about it in near future, just give him some time :)

**Comment by abhinavrastogi** [comment]
Waiting eagerly for this to land! Any ETA?

**Comment by lukekarrys** [comment]
Sorry for the delay in this. Now that the word is out re [converting to lodash](https://blog.andyet.com/2015/03/25/amped-down) I'm going to close this. I do think this PR will provide good reference for the upcoming conversion to lodash.

**Comment by latentflip** [comment]
Massive ^5 to @drewloomer for attempting this before we changed our minds on amp-core. It'll definitely be helpful for lodash conversion.

**Pull request closed without merging**